pkg/resizer/csi_resizer.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resizer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/kubernetes-csi/external-resizer/pkg/csi"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+)
+
+// NewCSIResizer creates a new resizer responsible for resizing CSI volumes.
+func NewCSIResizer(address string, timeout time.Duration) (Resizer, error) {
+	client, err := csi.New(address, timeout)
+	if err != nil {
+		return nil, fmt.Errorf("connect to CSI driver failed: %v", err)
+	}
+
+	if err := waitDriverReady(client, timeout); err != nil {
+		return nil, err
+	}
+
+	name, err := getDriverName(client, timeout)
+	if err != nil {
+		return nil, fmt.Errorf("get driver name failed: %v", err)
+	}
+
+	if err := supportsPluginControllerService(client, timeout); err != nil {
+		return nil, err
+	}
+
+	if err := supportsControllerResize(client, timeout); err != nil {
+		return nil, err
+	}
+
+	return &csiResizer{
+		name:    name,
+		client:  client,
+		timeout: timeout,
+	}, nil
+}
+
+type csiResizer struct {
+	name    string
+	client  csi.Client
+	timeout time.Duration
+}
+
+func (r *csiResizer) Name() string {
+	return r.name
+}
+
+func (r *csiResizer) CanSupport(pv *v1.PersistentVolume) bool {
+	source := pv.Spec.CSI
+	if source == nil {
+		klog.V(4).Infof("PV %s is not a CSI volume, skip it", pv.Name)
+		return false
+	}
+	if source.Driver != r.name {
+		klog.V(4).Infof("Skip resize PV %s for resizer %s", pv.Name, source.Driver)
+		return false
+	}
+	return true
+}
+
+func (r *csiResizer) Resize(pv *v1.PersistentVolume, requestSize resource.Quantity) (resource.Quantity, bool, error) {
+	oldSize := pv.Spec.Capacity[v1.ResourceStorage]
+
+	source := pv.Spec.CSI
+	if source == nil {
+		return oldSize, false, errors.New("not a CSI volume")
+	}
+	volumeID := source.VolumeHandle
+	if len(volumeID) == 0 {
+		return oldSize, false, errors.New("empty volume handle")
+	}
+
+	ctx, cancel := timeoutCtx(r.timeout)
+	defer cancel()
+	// TODO: Get secrets used for expansion operation.
```

Comment:
you can get the secret here and remove the TODO :)

Reply:
I'm not sure how to get these secrets. The external-attacher got publish/stage secrets from the PV object directly, and the external-snapshotter got snapshot secrets from `VolumeSnapshotClass.Parameters`. So which way should we adopt? Add a `ControllerExpandSecretRef` field to `CSIPersistentVolumeSource` or just fetch these secrets from `StorageClass.Parameters`?

cc @gnufied

Reply:
resizer needs admin credentials, not publish/stage credentials (used by attacher), in most cases.

Reply:
Lets do what `external-provisioner` does. It also fetches secrets from SC. We could use same secret that are being used for provisioning. cc @msau42

Reply:
Which secret needs the be used? Same as provision? Or do we need a new secret for resize?

Reply:
In most cases it will be same as provisioner secret but may be in some cases user have generated secrets which have permission to resize but not to provision? 

looks like `csi.storage.k8s.io/provisioner-secret-name/namespace` is what provisioner is using for fetching secrets. I think we will have to either introduce new parameters to SC or reuse provisioner secrets? Reusing provisioner secrets feel dirty so I guess the only workable long-term solution is to let SC define different secrets for resizing.

Reply:
I think having a separate secret for resize makes sense. Should it also be added to the CSI volume spec?

Reply:
you mean `CSIPersistentVolumeSource` ? I am not sure we need resizing secrets there because resizing is not supported for inline volume anyways.

Reply:
PV source

Reply:
PV source still refers to `CSIPersistentVolumeSource` right? So no - I do not think we need resizing secrets there.

Reply:
It's where we put all the other non provisioning secrets. So you don't need to reference storageclass.

Reply:
OK, I will fetch secrets from SC with a separate name.

Reply:
Instead of fetching secrets directly from the storageclass, can it be added as a parameter to the CSIPersistentVolumeSource?  That way all the secret fetching/templating logic can be confined to just the external-provisioner.